overlay: make overlay's style and colors configurable

tokyo4j opened the pull request:
This PR adds following theme settings:
- `snapping-preview.region.area.color`
- `snapping-preview.edge.area.color`
- `snapping-preview.region.rect-type` (`Auto|Area|Outlines`)
- `snapping-preview.edge.rect-type` (`Auto|Area|Outlines`)

I named the theme entry `rect-type` to make `labwc-theme.5.scd` easy to follow, but maybe it should be just `type`.

I restructured `struct overlay` to allow independent styles for region/edge snapping overlay.
I also added `inactivate_overlay()` as a refactoring.

In the future, maybe we can also make the colors of outlines configurable like:
`snapping-preview.region.outlines.colors: #ffffff,#000000,#ffffff`
In that case, I think the colors of OSD preview should also be configurable.

johanmalm commented:
Thanks. Looks good in general.

>     * `snapping-preview.region.area.color`
>     * `snapping-preview.edge.area.color`

For consistency with the rest of the `themerc` file I'd suggest `bg` instead of `area`.

`snapping-preview.bg.color`
`snapping-preview.border.color`

>     * `snapping-preview.region.rect-type` (`Auto|Area|Outlines`)
>     * `snapping-preview.edge.rect-type` (`Auto|Area|Outlines`)

> I named the theme entry `rect-type` to make `labwc-theme.5.scd` easy to follow, but maybe it should be just `type`.

I'm fine with `rect-type`. If you want to consider something else, we could go with `fill` (yes|no|auto). Fill is consistent with cairo-speak (fill/stroke).

For reference, `rc.xml` uses `<resize><drawContents>` - although that feels slightly different.

> In the future, maybe we can also make the colors of outlines configurable like: `snapping-preview.region.outlines.colors: #ffffff,#000000,#ffffff` In that case, I think the colors of OSD preview should also be configurable.

Good idea. Agree. Suggest calling it `border` rather than `outlines` for consistency.

tokyo4j commented:
I implemented @johanmalm's suggestions, as they make sense to me. `fill` sounds straightforward rather than defining `rect-type` type, as long as we only support `wlr_scene_rect` and `multi_rect`. If we may support another rectangle style in the future, that should be reverted.

I also implemented `snapping-preview.[region|edge].border.color`, whose possible values are two colors separated by a comma like `#ffffff,#000000` or `auto` keyword which falls back to OSD's background/text colors.

I haven't made OSD preview color configurable yet as it's not in the scope of this PR, but it should be an easy fix.

tokyo4j commented:
I reorganized my commits and some made changes in `snapping-preview.[region|edge].border.color`.

`snapping-preview.[region|edge].border.color` now allows specifying 1-3 colors, and only single line rectangle is drawn when only one color is specified.
Its default value is now set to `#ffffff,#000000,#ffffff` rather than falling back to OSD's colors.

I also added `snapping-preview.[region|edge].border.width`.

Consolatis commented on docs/labwc-theme.5.scd:
```diff
@@ -195,6 +195,52 @@ elements are not listed here, but are supported.
 
+*snapping-preview.region.fill* [yes|no|auto]
+	Show filled rectangle as preview when window is snapped to a region.
+	Otherwise, outlined rectangle is shown instead.
```
`Otherwise, an (a?) outlined rectangle is shown instead`.

Consolatis commented on docs/labwc-theme.5.scd:
```diff
@@ -195,6 +195,52 @@ elements are not listed here, but are supported.
+
+*snapping-preview.region.bg.color*
+	Color of filled rectangle shown as preview when window is snapped to a
+	region.
+	Default is #8080b380.
```
Default can be in the previous line, same for the following 3 entries. (I think it looks better when the the "Default is" line is shorter than the line above, in all other cases it should be in the previous line instead).

Consolatis commented on docs/labwc-theme.5.scd:
```diff
@@ -195,6 +195,52 @@ elements are not listed here, but are supported.
+
+*snapping-preview.region.border.color*
+	Color(s) of outlined rectangle shown as preview when window is snapped to a
+	region.
+	Possible value is a color or up to 3 colors separated by a comma
```
Similar thing here, I think the line break is not necessary to split the description and the value definition. Same for `snapping-preview.edge.border.color`.

Consolatis commented on docs/themerc:
```diff
@@ -71,3 +71,12 @@ osd.window-switcher.item.active.border.width: 2
+snapping-preview.edge.bg.color: #8080b380
+snapping-preview.region.border.width: 1
+snapping-preview.edge.border.width: 1
+snapping-preview.region.border.color: #ffffff,#000000,#ffffff
+snapping-preview.edge.border.color: #ffffff,#000000,#ffffff
```
Do we want to change the default to white-black-white here?
Using the OSD colors instead as default will somewhat match the theme in use and should also provide a high enough contrast ratio (as otherwise the text on the OSD wouldn't be readable).

tokyo4j replied:
That's right.
I didn't know that some OSD styles already falls back to values like `window.active.border.color` when I made this change, and thought that I need to allow any dedicated value like `auto` to make it fall back, which sounds tedious.

I will fix it soon.

tokyo4j commented:
Changes are:
- Adding a/an/the and minor fixes in the documentation
- Letting `snapping-preview.*.border.width` and `snapping-preview.*.border.color` fall back to OSD theme and fixing its documentation

tokyo4j commented:
I'm thinking of forcing outlines to be rendered when `WLR_RENDERER=pixman`.
I tried `snapping-preview.*.fill: yes` with `WLR_RENDERER=pixman` on my Ryzen 3 4300U laptop, and found it lags so badly (300ms~600ms in my sense).

So possible changes would be like:
- Possible values in `snapping-preview.*.fill` are just `yes|no`
- The default value for `snapping-preview.*.fill` is `yes`
- If the current rendering backend is pixman, outlines are always chosen

Consolatis commented:
Had a look through the diff again and didn't see anything obvious. 

- tested `Reconfigure` while a preview was shown: it just hides and everything works as expected thereafter
- tested `xcalc &; sleep 5; kill $(pidof xcalc)` and making xcalc show a preview: works as expected
- tested setting border width
- tested setting `.fill` to `no` on the gles2 renderer
- tested colors for border and area

Only slight nitpick I can think of:
`snapping-preview.` could be replaced by `snapping.preview.` (we had issues in the past where the `-` character used was actually some weird locale depended unicode character and not the ASCII version thus we failed to parse it correctly).

Other than that this looks great!
After addressing the `snapping-preview` vs `snapping.preview` point, do you want to squash yourself or should I do so when merging?

tokyo4j commented:
Thank you!

I'm fine with `snapping.preview`.

Honestly, I don't think the unicode issue (#1098?) can be a strong reason to choose it over `snapping-preview`, but `snapping.preview` looks more natural as there's `<snapping><preview>` in `rc.xml` so I like it better.

>  After addressing the snapping-preview vs snapping.preview point, do you want to squash yourself or should I do so when merging?

I'll do it myself.

Consolatis commented:
The last Last commit contains `snapping-overlay` in the commit body and the subject is 75 characters wide, suggest something like `overlay: add theme settings for colors and border width` instead. The commit body of the 2nd to last commit is slightly outdated.

tokyo4j commented:
Thank you! I've learned so much from you about how to contribute to OSS!

Consolatis commented:
Well, thank you for your great code and responsiveness to feedback. Really appreciated :)

I think the whole snapping overlay implementation is really nice out of the box but still flexible in case a user wants to change the behavior.

tokyo4j commented:
I'll also make the outline colors of window-switcher preview configurable in follow-up PRs.
It'll more likely to be worth changing from the default as its default colors are similar to that of window borders, which makes a bit harder to distinguish currently previewed window from others.
